src/lib/keygen.hpp

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright (c) 2024 [Ribose Inc](https://www.ribose.com).
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without modification,
+ * are permitted provided that the following conditions are met:
+ *
+ * 1.  Redistributions of source code must retain the above copyright notice,
+ *     this list of conditions and the following disclaimer.
+ *
+ * 2.  Redistributions in binary form must reproduce the above copyright notice,
+ *     this list of conditions and the following disclaimer in the documentation
+ *     and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef RNP_KEYGEN_HPP_
+#define RNP_KEYGEN_HPP_
+
+#include "repgp/repgp_def.h"
+#include "types.h"
+#include "sec_profile.hpp"
+#include "key_material.hpp"
+#include "pgp-key.h"
+
+namespace rnp {
+
+class KeygenParams {
+  private:
+    pgp_pubkey_alg_t                alg_;
+    pgp_hash_alg_t                  hash_;
+    pgp_version_t                   version_;
+    SecurityContext &               ctx_;
```

Comment:
Risky solution imho (to use reference here)

Reply:
SecurityContext is a singleton which lives the whole life of the top-level rnp_ffi_t object, and it would be destroyed after all the internal stuff. Is this good enough or there could be other solution?